File: src/test/java/fi/metatavu/oioi/cm/test/functional/tests/CustomerTestsIT.java
package fi.metatavu.oioi.cm.test.functional.tests;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertEquals;

import java.util.List;
import java.util.UUID;

import org.junit.Test;
import org.openapitools.client.model.Customer;

import fi.metatavu.oioi.cm.test.functional.builder.TestBuilder;

/**
 * Customer functional tests
 * 
 * @author Antti Leppä
 *
 */
public class CustomerTestsIT extends AbstractFunctionalTest {

  @Test
  public void testCreateCustomer() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      assertNotNull(builder.admin().customers().create("test customer", "http://example.com/great-image.jpg"));
    }
  }
  
  @Test
  public void testFindCustomer() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Customer createdCustomer = builder.admin().customers().create("test customer", "http://example.com/great-image.jpg");
      builder.admin().customers().assertFindFailStatus(404, UUID.randomUUID());
      Customer foundCustomer = builder.admin().customers().findCustomer(createdCustomer.getId());
      builder.admin().customers().assertCustomersEqual(createdCustomer, foundCustomer);
    }
  }
  
  @Test
  public void testListCustomers() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Customer createdCustomer = builder.admin().customers().create("test customer", "http://example.com/great-image.jpg");
      List<Customer> foundCustomers = builder.admin().customers().listCustomers();
      assertEquals(1, foundCustomers.size());
      builder.admin().customers().assertCustomersEqual(createdCustomer, foundCustomers.get(0));
    }
  }
  
  @Test
  public void testUpdateCustomer() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Customer createdCustomer = builder.admin().customers().create("test customer", "http://example.com/great-image.jpg");
      builder.admin().customers().assertCustomersEqual(createdCustomer, builder.admin().customers().findCustomer(createdCustomer.getId()));

      Customer updateCustomer = builder.admin().customers().findCustomer(createdCustomer.getId());
      updateCustomer.setName("updated customer");
      updateCustomer.setImageUrl("http://example.com/greater-image.jpg");
      Customer updatedCustomer = builder.admin().customers().updateCustomer(updateCustomer);
      assertEquals(createdCustomer.getId(), updatedCustomer.getId());
      assertEquals(updateCustomer.getName(), updatedCustomer.getName());
      assertEquals(updateCustomer.getImageUrl(), updatedCustomer.getImageUrl());
      Customer foundCustomer = builder.admin().customers().findCustomer(createdCustomer.getId());
      assertEquals(createdCustomer.getId(), foundCustomer.getId());
      assertEquals(updateCustomer.getName(), foundCustomer.getName());
      assertEquals(updateCustomer.getImageUrl(), foundCustomer.getImageUrl());
    }
  }

  @Test
  public void testDeleteCustomer() throws Exception {
    try (TestBuilder builder = new TestBuilder()) {
      Customer createdCustomer = builder.admin().customers().create("test customer", "http://example.com/great-image.jpg");
      Customer foundCustomer = builder.admin().customers().findCustomer(createdCustomer.getId());
      assertEquals(createdCustomer.getId(), foundCustomer.getId());
      builder.admin().customers().delete(createdCustomer);
      builder.admin().customers().assertDeleteFailStatus(404, createdCustomer);
    }
  }
  
}

File: src/test/kotlin/fi/metatavu/oioi/cm/test/functional/tests/CustomerTestsIT.kt
package fi.metatavu.oioi.cm.test.functional.tests

import fi.metatavu.oioi.cm.test.functional.resources.MysqlResource
import fi.metatavu.oioi.cm.client.models.Customer

import kotlin.Throws
import fi.metatavu.oioi.cm.test.functional.builder.TestBuilder
import fi.metatavu.oioi.cm.test.functional.resources.HiveMQTestResource
import fi.metatavu.oioi.cm.test.functional.resources.KeycloakTestResource
import io.quarkus.test.common.QuarkusTestResource
import io.quarkus.test.junit.QuarkusTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.lang.Exception
import java.util.UUID

/**
 * Customer functional tests
 *
 * @author Antti Leppä
 */
@QuarkusTest
@QuarkusTestResource.List(
    QuarkusTestResource(KeycloakTestResource::class),
    QuarkusTestResource(MysqlResource::class),
    QuarkusTestResource(HiveMQTestResource::class)
)
class CustomerTestsIT : AbstractFunctionalTest() {
    @Test
    @Throws(Exception::class)
    fun testCreateCustomer() {
        TestBuilder().use { builder ->
            assertNotNull(
                builder.admin.customers.create(
                    "test customer",
                    "http://example.com/great-image.jpg"
                )
            )
        }
    }

    @Test
    @Throws(Exception::class)
    fun testFindCustomer() {
        TestBuilder().use { builder ->
            val createdCustomer = builder.admin.customers.create("test customer", "http://example.com/great-image.jpg")
            builder.admin.customers.assertFindFailStatus(404, UUID.randomUUID())
            val foundCustomer = builder.admin.customers.findCustomer(createdCustomer.id)
            builder.admin.customers.assertCustomersEqual(createdCustomer, foundCustomer)
        }
    }

    @Test
    @Throws(Exception::class)
    fun testListCustomers() {
        TestBuilder().use { builder ->
            val createdCustomer = builder.admin.customers.create("test customer", "http://example.com/great-image.jpg")
            val foundCustomers: Array<Customer> = builder.admin.customers.listCustomers()
            assertEquals(1, foundCustomers.size.toLong())
            builder.admin.customers.assertCustomersEqual(createdCustomer, foundCustomers[0])
        }
    }

    @Test
    @Throws(Exception::class)
    fun testListCustomerPermissions() {
        TestBuilder().use { builder ->
            builder.admin.customers.create("customer-1", "http://example.com/great-image.jpg")
            builder.admin.customers.assertCount(1)
            builder.customer1User.customers.assertCount(1)
            builder.customer1Admin.customers.assertCount(1)
            builder.customer2User.customers.assertCount(0)
            builder.customer2Admin.customers.assertCount(0)
        }
    }

    @Test
    @Throws(Exception::class)
    fun testUpdateCustomer() {
        TestBuilder().use { builder ->
            val createdCustomer = builder.admin.customers.create("test customer", "http://example.com/great-image.jpg")

            builder.admin.customers.assertCustomersEqual(
                createdCustomer,
                builder.admin.customers.findCustomer(createdCustomer.id)
            )

            val updateCustomer = builder.admin.customers.findCustomer(createdCustomer.id).copy(
                name = "updated customer",
                imageUrl = "http://example.com/greater-image.jpg"
            )

            val (name, id, imageUrl) = builder.admin.customers.updateCustomer(updateCustomer)
            assertEquals(createdCustomer.id, id)
            assertEquals(updateCustomer.name, name)
            assertEquals(updateCustomer.imageUrl, imageUrl)
            val (name1, id1, imageUrl1) = builder.admin.customers.findCustomer(createdCustomer.id)
            assertEquals(createdCustomer.id, id1)
            assertEquals(updateCustomer.name, name1)
            assertEquals(updateCustomer.imageUrl, imageUrl1)
        }
    }

    @Test
    @Throws(Exception::class)
    fun testDeleteCustomer() {
        TestBuilder().use { builder ->
            val createdCustomer =
                builder.admin.customers.create("test customer", "http://example.com/great-image.jpg")
            val (_, id) = builder.admin.customers.findCustomer(createdCustomer.id)
            assertEquals(createdCustomer.id, id)
            builder.admin.customers.delete(createdCustomer)
            builder.admin.customers.assertDeleteFailStatus(404, createdCustomer)
        }
    }

    @Test
    fun testDeletePermissions() {
        TestBuilder().use { builder ->
            val customer = builder.admin.customers.create(name = "customer-1")

            builder.customer2User.customers.assertDeleteFailStatus(
                expectedStatus = 403,
                customer = customer
            )

            builder.customer2Admin.customers.assertDeleteFailStatus(
                expectedStatus = 403,
                customer = customer
            )

            builder.customer1User.customers.assertDeleteFailStatus(
                expectedStatus = 403,
                customer = customer
            )

            builder.customer1Admin.customers.assertDeleteFailStatus(
                expectedStatus = 403,
                customer = customer
            )

            builder.admin.customers.delete(
                customer = customer
            )
        }
    }
}